src/db/gc/mod.rs

```diff
@@ -0,0 +1,307 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+//!
+//! The current implementation of the garbage collector is `concurrent mark-and-sweep`.
+//!
+//! ## Terminology
+//! `chain finality` - a number of epochs after which it becomes impossible to add or remove a block
+//! previously appended to the blockchain.
+//!
+//! ## Design goals
+//! A correct GC algorithm that is simple and efficient for forest scenarios.
+//!
+//! ## GC Algorithm
+//! The `mark-and-sweep` algorithm was chosen due to it's simplicity, efficiency and low memory
+//! footprint. Previously the `semi-space` algorithm was used resulting in data duplication and up
+//! to a 100% extra disk usage.
+//!
+//! ## GC Workflow
+//! 1. Mark: traverse all the relevant database columns, generating integer hash representations for
+//! each database key and storing those in a set.
+//! 2. Wait at least `chain finality` blocks.
+//! 3. Traverse reachable blocks starting at the current heaviest tipset and remove those from the
+//! marked set, leaving only unreachable entries that are older than `chain finality` to avoid
+//! removing something that could later become reachable as a result of a fork.
```

Comment:
> leaving only unreachable entries that are older than `chain finality` to avoid removing something that could later become reachable as a result of a fork.

This is important information. I feel this should go into the design goals as part of the problem description.